src/Channels.Networking.Sockets/SocketConnection.cs

```diff
@@ -0,0 +1,304 @@
+﻿using System;
+using System.Net;
+using System.Net.Sockets;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Channels.Networking.Sockets
+{
+    public class SocketConnection : IChannel
+    {
+        private Socket _socket;
+        private Channel _input, _output;
+        private ChannelFactory _channelFactory;
+        private readonly bool _ownsChannelFactory;
+
+        static readonly EventHandler<SocketAsyncEventArgs> _asyncCompleted = OnAsyncCompleted;
+
+        public IReadableChannel Input => _input ?? CreateInput();
+        public IWritableChannel Output => _output ?? CreateOutput();
+
+        private Channel CreateInput()
+        {
+            _input = ChannelFactory.CreateChannel();
+            ProcessReads();
+            return _input;
+        }
+
+        private Channel CreateOutput()
+        {
+            _output = ChannelFactory.CreateChannel();
+            ProcessWrites();
+            return _output;
+        }
+
+        private Socket Socket => _socket;
+        private ChannelFactory ChannelFactory => _channelFactory;
+
+        internal SocketConnection(Socket socket, ChannelFactory channelFactory)
```

Comment:
Would like to continue to discuss and evolve this with you independently to the PR; right now, this sounds a lot like `SocketClientConnectionFactory` or something equally java-esque. I agree that there is some over-arching "scope" thing missing here, but it would be good to have a discussion about what this *should be* rather than forcing something in for the sake of it. At the moment the libuv stuff gets away with this by hiding it behind the `UvThread`, but conceptually I'd argue that the thread is **not** the same thing (scope-wise and meaning-wise) as the thing we're talking about here.